src/Web/Rsp.IrasPortal.Web/Controllers/QuestionSetController.cs

```diff
@@ -1,287 +1,215 @@
+        if (response.IsSuccessStatusCode)
+        {
+            model.Versions = response.Content?.OrderByDescending(x => x.CreatedAt).ToList() ?? [];
+        }
     }
 
     [HttpPost]
-    public async Task<IActionResult> Upload(QuestionSetFileModel model)
+    public async Task<IActionResult> Upload(QuestionSetViewModel model)
     {
         var file = model.Upload;
+        await GetVersions(model);
 
         if (file == null || file.Length == 0)
         {
             ModelState.AddModelError("Upload", "Please upload a file");
-            return View(model);
+            return View(nameof(Index), model);
         }
 
-        string[] allowedExtensions = [".xlsx", ".xlsb", ".xls"];
-
-        if (!allowedExtensions.Contains(Path.GetExtension(file.FileName).ToLower()))
+        if (model.Versions.Any(v => v.VersionId
+                .Equals(Path.GetFileNameWithoutExtension(file.FileName), StringComparison.CurrentCultureIgnoreCase)))
         {
-            ModelState.AddModelError("Upload", $"Please upload a file of type: {string.Join(", ", allowedExtensions)}");
-            return View(model);
+            ModelState.AddModelError("Upload", "Version name already exists");
+            return View(nameof(Index), model);
         }
 
-        Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
-        using var stream = file.OpenReadStream();
-        using var reader = ExcelReaderFactory.CreateReader(stream);
-        var result = reader.AsDataSet(new ExcelDataSetConfiguration()
-        {
-            ConfigureDataTable = (_) => new ExcelDataTableConfiguration()
-            {
-                UseHeaderRow = true
-            }
-        });
+        var fileProcessResponse = questionSetService.ProcessQuestionSetFile(file);
 
-        List<DataTable> moduleSheets = [];
-        DataTable rulesSheet = new();
-        DataTable answerOptionsSheet = new();
-        DataTable contentsSheet = new();
-
-        foreach (var sheetName in SheetNames.All)
+        if (fileProcessResponse.Error != null)
         {
-            var sheet = result.Tables[sheetName];
-
-            if (sheet == null)
-            {
-                ModelState.AddModelError("Upload", $"Sheet '{sheetName}' cannot be found");
-                continue;
-            }
-
-            if (HasValidColumns(sheetName, sheet))
-            {
-                if (sheetName == SheetNames.Rules)
-                {
-                    rulesSheet = sheet;
-                }
-                else if (sheetName == SheetNames.AnswerOptions)
-                {
-                    answerOptionsSheet = sheet;
-                }
-                else if (sheetName == SheetNames.Contents)
-                {
-                    contentsSheet = sheet;
-                }
-                else
-                {
-                    moduleSheets.Add(sheet);
-                }
-            }
+            ModelState.AddModelError("Upload", fileProcessResponse.ReasonPhrase!);
```

Comment:
use nameof(Upload) instead of "Upload"
```suggestion
            ModelState.AddModelError(nameof(Upload), fileProcessResponse.ReasonPhrase!);
```